### PR TITLE
End to end test for Telemetry.

### PR DESCRIPTION
**Pull request by khilangoel**

### CODE REVIEW

File: test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        private readonly string resultPath;
+        private const string TELEMETRY_OPTEDIN = "VSTEST_TELEMETRY_OPTEDIN";
+        private const string LOG_TELEMETRY = "VSTEST_LOGTELEMETRY";
+
+        public TelemetryTests()
+        {
+            this.resultPath = Path.GetTempPath() + "TelemetryLogs";
+
+            // Opt in the Telemetry
+            Environment.SetEnvironmentVariable(TELEMETRY_OPTEDIN, "1");
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs]: Yes, that makes sense. will do it,

File: src/vstest.console/Publisher/MetricsPublisher.cs [hunk 2]
```diff
@@ -134,5 +138,31 @@ public void Dispose()
...
+        {
+            var logEnabled = Environment.GetEnvironmentVariable("VSTEST_LOGTELEMETRY");
+
+            if (!string.IsNullOrEmpty(logEnabled) && logEnabled.Equals("1", StringComparison.Ordinal))
+            {
+                string resultDirectory = Path.GetTempPath() + "TelemetryLogs";
+                string resultFileName = Guid.NewGuid().ToString();
+                string path = Path.Combine(resultDirectory, resultFileName);
+
+                if (!Directory.Exists(resultDirectory))
```
[the remark this answers is not in the document]
Comment (khilan2goel) [reply src/vstest.console/Publisher/MetricsPublisher.cs]: Done

File: src/vstest.console/Publisher/MetricsPublisher.cs [hunk 3]
```diff
@@ -70,6 +75,13 @@ public void PublishMetrics(string eventName, IDictionary<string, object> metrics
                 }
 
                 this.session.PostEvent(telemetryEvent);
+
+                // Log to Text File
+                var logEnabled = Environment.GetEnvironmentVariable("VSTEST_LOGTELEMETRY");
```
Comment (codito) [inline src/vstest.console/Publisher/MetricsPublisher.cs]: Can you add the telemetry key/value pairs in `EqtTrace` on line 63?

File: test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            {
+                var directory = new DirectoryInfo(this.resultPath);
+                var file = directory.GetFiles().OrderByDescending(f => f.CreationTime).First();
+
+                string[] lines = File.ReadAllLines(file.FullName);
+                foreach (var line in lines)
+                {
+                    if (line.Contains(TelemetryDataConstants.TestExecutionCompleteEvent))
+                    {
+                        var isPresent = line.Contains(TelemetryDataConstants.DataCollectorsEnabled)
```
Comment (codito) [inline test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs]: Do we need to validate the correctness of values?

File: test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+            if (Directory.Exists(this.resultPath))
+            {
+                var directory = new DirectoryInfo(this.resultPath);
+                var file = directory.GetFiles().OrderByDescending(f => f.CreationTime).First();
+
+                string[] lines = File.ReadAllLines(file.FullName);
+                foreach (var line in lines)
+                {
+                    if (line.Contains(TelemetryDataConstants.TestExecutionCompleteEvent))
```
Comment (codito) [inline test/Microsoft.TestPlatform.AcceptanceTests/TelemetryTests.cs]: Shouldn't the `Execution` or `Discovery` come in as an argument to the assert?

**Review by codito — APPROVED** [review]
Added few suggestions

File: src/vstest.console/Publisher/MetricsPublisher.cs (hunk 3 above)
Comment (khilan2goel) [reply src/vstest.console/Publisher/MetricsPublisher.cs]: Done.

**Comment by khilan2goel** [comment]
@dotnet-bot  test Windows_NT / Debug Build

**Review by harshjain2 — APPROVED** [review]

**Comment by khilan2goel** [comment]
@dotnet-bot test Windows_NT / Release Build

**Comment by harshjain2** [comment]
@dotnet-bot test Windows_NT / Debug Build please.

**Pull request merged**